### PR TITLE
Add configuration for black autoformatter

### LINKED ISSUE #13942: Create `black` configuration
Turn the `black` configuration described in [APE 20](https://github.com/astropy/astropy-APEs/blob/main/APE20.rst) into a configuration for `astropy` which can be then implemented progressively as described.

### PR DESCRIPTION
**Pull request by WilliamJamieson**
### Description

This PR introduces a black configuration for astropy that allows each subpackage to opt-in to using the black autoformatter similar to #12932. This is a draft of the implementation [discussed](https://docs.google.com/document/d/15JSFh3OMF9Iz6ov3q_xxGO_BL8hRnuse4IMUrqEIvcg/edit#bookmark=id.vjp5cvt4xv36) during the recent astropy dev telecon.

Approval Matrix:

| sub-package | maintainer | approved |
| --- | --- | --- |
| convolution | @larrybradley | ✅ |
| coordinates | |  ❌ |
| cosmology| @nstarman | ✅  |
| io |  | ❌ |
| io.ascii | @taldcroft  | ✅ |
| io.fits | | ❌ |
| io.misc | | ❌ |
| io.registry | @nstarman | ✅  |
| io.votable | | ❌ |
| modeling | @nden | ✅ |
| nddata |  | ❌ |
| samp |  | ❌ |
| stats | @larrybradley | ✅ |
| table | @taldcroft | ✅ |
| time | @taldcroft  | ✅ |
| timeseries | @astrofrog  | ✅ |
| units | @nstarman | ✅  |
| utils |  | ❌ |
| visualization | @astrofrog | ✅ |
| wcs | @mcara | ✅ |

resolves #13942

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Do the proposed changes actually accomplish desired goals?
- [x] Do the proposed changes follow the [Astropy coding guidelines](https://docs.astropy.org/en/latest/development/codeguide.html)?
- [x] Are tests added/updated as required? If so, do they follow the [Astropy testing guidelines](https://docs.astropy.org/en/latest/development/testguide.html)?
- [x] Are docs added/updated as required? If so, do they follow the [Astropy documentation guidelines](https://docs.astropy.org/en/latest/development/docguide.html#astropy-documentation-rules-and-guidelines)?
- [x] Is rebase and/or squash necessary? If so, please provide the author with appropriate instructions. Also see ["When to rebase and squash commits"](https://docs.astropy.org/en/latest/development/when_to_rebase.html).
- [x] Did the CI pass? If no, are the failures related? If you need to run daily and weekly cron jobs as part of the PR, please apply the `Extra CI` label.
- [x] Is a change log needed? If yes, did the change log check pass? If no, add the `no-changelog-entry-needed` label. If this is a manual backport, use the `skip-changelog-checks` label unless special changelog handling is necessary.
- [x] Is this a big PR that makes a "What's new?" entry worthwhile and if so, is (1) a "what's new" entry included in this PR and (2) the "whatsnew-needed" label applied?
- [x] Is a milestone set? Milestone must be set but `astropy-bot` check might be missing; do not let the green checkmark fool you.
- [x] At the time of adding the milestone, if the milestone set requires a backport to release branch(es), apply the appropriate `backport-X.Y.x` label(s) *before* merge.

### CODE REVIEW

File: pyproject.toml
Comment (WilliamJamieson) [inline pyproject.toml:50]: @nstarman I had to remove this part of the `isort` configuration in order to prevent `black` from fighting each other. Is there an alternate way to handle this?

Comment (WilliamJamieson) [reply pyproject.toml:50]: This https://github.com/psf/black/issues/127#issuecomment-381032478 indicates that `multi_line_output` mode `8` is the output mode (which is the mode set by the `profile = "black"`) that allows `isort` and `black` to be compatible.

Comment (nstarman) [reply pyproject.toml:50]: Unfortunately there is no alternative. ``isort`` supports styles that black does not. We had ``multi_line_output = 4`` because it was similar to Astropy's current standard. There is another ``multi_line_output`` mode that could work: **9** in https://pycqa.github.io/isort/docs/configuration/multi_line_output_modes.html

Comment (nstarman) [reply pyproject.toml:50]: Not sure it's any better...

Comment (nstarman) [reply pyproject.toml:50]: Just fewer lines

Comment (WilliamJamieson) [reply pyproject.toml:50]: Here is the difference between the two settings: https://github.com/WilliamJamieson/astropy/compare/black-config..black-config-alternate-multiline.

I personally prefer the current (mode 8) as I think all the additional boiler-plate seems excessive. Thoughts?

Comment (nstarman) [reply pyproject.toml:50]: Agreed, though it's one of the few things I don't enjoy about black.

**Comment by eteq** [comment]
I am deeply, deeply against ever applying Black Astropy.  In my opinion it is a deep and heinous violation of PEP8 which begins with: "Foolish Consistency is the Hobgoblin of Little Minds" and basically says "don't blindly apply style guides, instead think about code readability".

Now the counter-argument to this is "I don't want to think about it."  But to that I say "I'm sorry to hear that, but you have to." Readability Counts is a core principle of Python.

**Comment by eteq** [comment]
And to be crystal-clear, I realize I don't get a veto here if the will of the majority is to start doing Black.  But I do get to argue against it and try to persuade others :wink:

**Comment by karllark** [comment]
As a very minor bit player in astropy, I would advocate for using black.  I have been using it on most the packages I maintain including `dust_extinction` and `beast`.   It has been great for me to avoid having to deal with PEP8 issues *and* has been great for new contributors to the `beast` as learning how to write well formatted python code is *not* trivial.

Not having to worry about the formatting has removed stress from my life.

I note that I do occasionally turn off the black autoformatting using the comments `# fmt: off` and  `# fmt: on` where black does make things egregious.

File: pyproject.toml
Comment (nstarman) [inline pyproject.toml:36]: include cosmology! Or do you want to do application in followup PRs?

**Comment by taldcroft** [comment]
From people that have expressed an opinion on the astropy-dev thread, it seems that most favor either (1) applying black to the entire astropy core or (2) don't do black at all. So unless we hear from people that they would be fine with black but not in their subpackage, then I'd suggest this PR be simplified to apply black to the whole package.

File: pyproject.toml
Comment (WilliamJamieson) [reply pyproject.toml:36]: I think doing them in followup PRs would be better, unless we end up deciding everything should be blacked. In which case everything should be in one giant change?

Comment (nstarman) [reply pyproject.toml:36]: It might be easier to apply ``black`` globally by doing it sequentially to supbackages.  Merge conflicts abound while waiting for subpackage maintainer approval.

**Comment by nstarman** [comment]
How does this PR factor into the adoption of ``black``. Is this the branch to review?

**Comment by WilliamJamieson** [comment]
> How does this PR factor into the adoption of `black`. Is this the branch to review?

@taldcroft knows more than I do, but I believe this PR would be the first step in implementing `black`. It should layout all of the configuration options for `black` it only suppresses actually applying `black`. So feel free to start reviewing this in detail.

File: pyproject.toml
Comment (nstarman) [reply pyproject.toml:54]: Well, black is in so this discussion is moot.

File: astropy/units/physical.py
Comment (nstarman) [inline astropy/units/physical.py:14]: Can you move the comment back to `imperial`?

Comment (WilliamJamieson) [reply astropy/units/physical.py]: Done

File: astropy/units/physical.py
Comment (WilliamJamieson) [inline astropy/units/physical.py:10]: @nstarman why is `yesqa` stripping the `# noqa` off this line?

File: pyproject.toml
Comment (nstarman) [reply pyproject.toml:62]: Close now that black APE is accepted?

File: astropy/units/physical.py
Comment (nstarman) [inline astropy/units/physical.py:11]:
```suggestion
from . import imperial  # noqa: F401  # 4 bkwd compat  #11975 and #11977
```

File: astropy/units/physical.py
Comment (nstarman) [reply astropy/units/physical.py]: Hmm.  maybe because F401 is not in https://github.com/astropy/astropy/blob/3132229f01427ee0637c038c7475cf8b6af6d8ee/.pre-commit-config.yaml#L116

Comment (nstarman) [reply astropy/units/physical.py]: Or move the noqa comment to the end

Comment (nstarman) [reply astropy/units/physical.py]: maybe similar to https://stackoverflow.com/questions/51179109/set-pyflake-and-mypy-ignore-same-line

Comment (WilliamJamieson) [reply astropy/units/physical.py]: It's not a fan of having the comment and noqa comment on the same line.

Comment (nstarman) [reply astropy/units/physical.py]: Any way that works. 👍

**Comment by WilliamJamieson** [comment]
Are the 3.11 failures related to this PR?

**Comment by pllim** [comment]
I haven't seen that particular failure anywhere else, so probably?

**Comment by pllim** [comment]
But I can't see anything obvious in the diff on why it would break. Maybe wait a few days and see if the same failures bubble up in the cron jobs.

Also when this is ready, slap the "Extra CI" label on it to make sure "exotic" jobs also pass before merge.

p.s. That comma before `)` really annoys me...

**Comment by pllim** [comment]
Update: I just saw https://github.com/astropy/astropy/issues/13873 so the failure isn't related to this PR.

**Comment by WilliamJamieson** [comment]
> Also when this is ready, slap the "Extra CI" label on it to make sure "exotic" jobs also pass before merge.

This PR is essentially ready. Should I go ahead and take it out of draft form?

File: pyproject.toml
Comment (saimn) [inline pyproject.toml:48]: I think we agreed on 88 ? (flake8 is set to 88 below, what's the default for isort ?)

**Comment by saimn** [comment]
I think you should create the `.git-blame-ignore-revs` file here too, so formatting commits can be added to it in each PR ?

**Comment by WilliamJamieson** [comment]
> I think you should create the `.git-blame-ignore-revs` file here too, so formatting commits can be added to it in each PR ?

I suppose this PR should add its `isort` changes to this before I merge

File: pyproject.toml
Comment (saimn) [inline pyproject.toml:104]: `use_parentheses` can be removed.
Why do we need `indented_import_headings` ?

Comment (WilliamJamieson) [reply pyproject.toml:104]: @nstarman do you know why?

**Comment by WilliamJamieson** [comment]
> That was also my understanding of the majority decision, that black is to be applied to the entire package (except for possibly some very specific pieces of code). So is there still anything really to opt-in now?

Correct, [APE 20](https://github.com/astropy/astropy-APEs/pull/76) accepted `black` for the entire core `astropy` library. However, its plan for implementing these changes in `astropy` involves doing a PR for each sub-package applying `black` so that reviewing the changes is more manageable. This PR is just setting up for each of those follow-on PRs, that is the per-sub-package settings are just in place temporarily so the CI works for each of those PRs.

File: pyproject.toml
Comment (WilliamJamieson) [reply pyproject.toml:104]: Removing `indented_import_headings` setting seemed to have no effect when I ran `isort` so I will remove it.

**Comment by dhomeier** [comment]
So we really should review and comment on the follow-up PRs for our respective sub-packages?

**Comment by WilliamJamieson** [comment]
> So we really should review and comment on the follow-up PRs for our respective sub-packages?

Yes, this is an opportunity to review the "code readability" of each package.

File: pyproject.toml
Comment (nstarman) [reply pyproject.toml]: > do you know why?

no, I forget! If it has no effect, then we're good :)

**Comment by taldcroft** [comment]
@WilliamJamieson - this looks generally good. I checked that the configuration updates make sense to me and that all the changes in Python code files within astropy/ are just import reformatting.

I think the key big-picture thing missing here is [bullet (1) of the Implementation](https://github.com/astropy/astropy-APEs/blob/main/APE20.rst#implementation):

Updating the [coding style](https://docs.astropy.org/en/latest/development/codeguide.html#coding-style-conventions):

- Adding documentation about Black, with a particular emphasis on the rationale expressed here and the exception policy described above.
- Document the Black exception guidelines (as described in this APE)
- Removing other references to code formatting in the documentation.

It might actually make sense to do that as a separate PR to make the review more manageable. We can think of that as another "sub-package" that needs Black applied and put that into the Approval Matrix in the Description above.

**Comment by taldcroft** [comment]
There is a conflict with `.pre-commit-config.yaml`.

**Comment by taldcroft** [comment]
Another point is that the initial Description is a bit outdated. It should reference APE 20 and not refer to the PR as a draft.

We probably want to separate this PR from all the subsequent PR's for each sub-package. I think it might make sense to have either one master issue which is the end-to-end Black process with checkboxes for each sub-package, OR make a new GitHub project page for this with each of the sub-package PR's linked. @nstarman - you seem pretty good with projects!

**Comment by WilliamJamieson** [comment]
Why are the docs failing now?

File: astropy/coordinates/tests/test_celestial_transformations.py
Comment (astrofrog) [inline astropy/coordinates/tests/test_celestial_transformations.py:22]: Just curious, is it normal that these are not sorted alphabetically so that eg FK4 is just before FK4NoETerms?

Comment (WilliamJamieson) [reply astropy/coordinates/tests/test_celestial_transformations.py:22]: @nstarman do you know why?

**Review by astrofrog — APPROVED** [review]
timeseries and visualization look good to me

File: astropy/coordinates/tests/test_celestial_transformations.py
Comment (taldcroft) [reply astropy/coordinates/tests/test_celestial_transformations.py:22]: I also noticed another unusual sorting:
```
    SIP,
    Chebyshev1D,
    Chebyshev2D,
    Hermite1D,
   ...
```

**Comment by dhomeier** [comment]
`io.ascii` looks hardly worse than expected to me, too.

File: astropy/coordinates/tests/test_celestial_transformations.py
Comment (nstarman) [reply astropy/coordinates/tests/test_celestial_transformations.py:22]: Hm. `isort` puts classes before functions. I would have guessed that it's getting confused and thinking ``FK4NoETerms`` is a function because of the ``format_doc`` decorator, but ``FK4`` has that as well.

Comment (WilliamJamieson) [reply astropy/coordinates/tests/test_celestial_transformations.py:22]: `modeling` also appears to suffer from this problem...

Comment (nstarman) [reply astropy/coordinates/tests/test_celestial_transformations.py:22]: Sans manual digging, that was my one and only guess. Pragmatically, if it's consistent between applications of ``pre-commit run -a`` then I'm inclined to continue with this PR. If we can understand and fix the sort, even better, but if not, when someone has the spare hours and the inclination we can re-sort and git blame ignore a fix at a later date.

**Comment by WilliamJamieson** [comment]
I don't think the docs build failure is due to this PR:
```
encountered some issues with some of the inventories, but they had working alternatives:
intersphinx inventory 'https://docs.scipy.org/doc/scipy/reference/objects.inv' not fetchable due to <class 'requests.exceptions.ConnectionError'>: HTTPSConnectionPool(host='docs.scipy.org', port=443): Max retries exceeded with url: /doc/scipy/reference/objects.inv (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x7fb6c9955700>: Failed to establish a new connection: [Errno 111] Connection refused'))
```

File: .pre-commit-config.yaml
Comment (nstarman) [inline .pre-commit-config.yaml:59]: We may be bumping to py3.9+ soon.

**Review by nstarman — APPROVED** [review]
Cosmology, io/registry, and units LGTM!

File: .pre-commit-config.yaml
Comment (WilliamJamieson) [reply .pre-commit-config.yaml:59]: I just moved it to be before `isort` and `black` so that one should only have to run pre-commit once. In my experience its change sometimes get reformatted slightly by `black` in particular).

**Review by nden — APPROVED** [review]
Changes in modeling look good.

**Review by mcara — APPROVED** [review]
`wcs` looks OK to me.

**Review by larrybradley — APPROVED** [review]
convolution, stats, and visualization LGTM.

**Review by taldcroft — APPROVED** [review]
LGTM

**Review by saimn — APPROVED** [review]
LGTM

**Pull request merged**
